### PR TITLE
Nb/feat/api round 2

### PR DESCRIPTION
**Pull request by Nadav-Barak**
Support tensorflow dataset.
Added coco dataset that is based on tensorflow
Fixed bug in properties outlier when there is a batch with no labels for a good reason
Fixed bug in properties outlier to display all boxes for relevant images under num_of_bboxes property
Added seed to *all* vision tests
Resolves https://github.com/deepchecks/deepchecks/issues/2218

### CODE REVIEW

File: deepchecks/vision/checks/data_integrity/abstract_property_outliers.py [hunk 1]
```diff
@@ -196,12 +193,14 @@ def _get_property_outlier_images(self, prop_name: str, lower_limit: float, upper
         """Get outlier images and their values for provided property."""
         result = []
         for idx, value in enumerate(self._lowest_property_value_images[prop_name]['property_values']):
+            value = value[0] if isinstance(value, t.Sequence) else value  # case of value per label value is a list
```
Comment (JKL98ISR) [inline deepchecks/vision/checks/data_integrity/abstract_property_outliers.py:196]: what

File: deepchecks/vision/checks/data_integrity/abstract_property_outliers.py [hunk 2]
```diff
@@ -109,12 +109,9 @@ def update(self, context: Context, batch: BatchWrapper, dataset_kind: DatasetKin
                 labels = [[label_per_image] for label_per_image in batch.numpy_labels]
             else:
                 labels = batch.numpy_labels
-            if not isinstance(property_values[0], t.Sequence):  # TODO: change after changing properties api
-                property_values = [[x] for x in property_values]
 
-            self._properties_results[prop_name].extend(property_values)
             self._images_uuid += batch.numpy_image_identifiers
-            self._update_lowest_highest_property_values(batch.numpy_images, labels, property_values, prop_name)
+            self._cache_property_values_and_images(batch.numpy_images, labels, list(property_values), prop_name)
```
Comment (JKL98ISR) [inline deepchecks/vision/checks/data_integrity/abstract_property_outliers.py:114]: `list(`
why you care? is sequence not enough

File: deepchecks/vision/context.py [hunk 3]
```diff
@@ -44,17 +44,16 @@ def __init__(
...
             raise DatasetValidationError('Can\'t initialize context with only test. if you have single dataset, '
                                          'initialize it as train')
-        if train and test:
-            validate_vision_data_compatibility(train, test)
 
-        if train is not None:
-            train.init_cache()
-            train.name = DEFAULT_DATASET_NAMES[0] if train.name is None else train.name
-            self._task_type = train.task_type
+        train.init_cache()
```
Comment (JKL98ISR) [inline deepchecks/vision/context.py:51]: train is never none?

File: deepchecks/vision/datasets/classification/mnist.py [hunk 4]
```diff
@@ -205,9 +215,111 @@ def __call__(self, batch):
...
+        if not self._check_exists():
+            raise RuntimeError('Dataset not found.' +
+                               ' You can use download=True to download it')
+
+        self.data, self.targets = self._load_data()
+        if n_samples:
+            self.data = self.data[:n_samples]
+            self.targets = self.targets[:n_samples]
+        if self.transform is not None:
+            self.data = torch.stack([self.transform(image=img.numpy())['image'] for img in self.data])
```
Comment (JKL98ISR) [inline deepchecks/vision/datasets/classification/mnist.py:262]: why not .cpu.detach?

File: deepchecks/vision/datasets/classification/mnist.py [hunk 5]
```diff
@@ -205,9 +215,111 @@ def __call__(self, batch):
...
+                    download_and_extract_archive(
+                        url, download_root=self.raw_folder,
+                        filename=filename,
+                        md5=md5
+                    )
+                except URLError as error:
+                    print(f'Failed to download (trying next):\n{error}')
+                    continue
+                finally:
+                    print()
```
Comment (JKL98ISR) [inline deepchecks/vision/datasets/classification/mnist.py:309]: what

File: deepchecks/vision/datasets/classification/mnist.py [hunk 6]
```diff
@@ -205,9 +215,111 @@ def __call__(self, batch):
...
+            for mirror in self.mirrors:
+                url = f'{mirror}{filename}'
+                try:
+                    print(f'Downloading {url}')
+                    download_and_extract_archive(
+                        url, download_root=self.raw_folder,
+                        filename=filename,
+                        md5=md5
+                    )
+                except URLError as error:
```
Comment (JKL98ISR) [inline deepchecks/vision/datasets/classification/mnist.py:305]: just print? use logger

File: deepchecks/vision/datasets/classification/mnist.py (hunk 4 above)
Comment (JKL98ISR) [reply deepchecks/vision/datasets/classification/mnist.py:262]: use object_to_numpy

File: deepchecks/vision/checks/data_integrity/abstract_property_outliers.py (hunk 2 above)
Comment (Nadav-Barak) [reply deepchecks/vision/checks/data_integrity/abstract_property_outliers.py:114]: its for user defined custom properties

File: deepchecks/vision/context.py (hunk 3 above)
Comment (Nadav-Barak) [reply deepchecks/vision/context.py:51]: true, if none its failing above

File: deepchecks/vision/datasets/classification/mnist.py (hunk 4 above)
Comment (Nadav-Barak) [reply deepchecks/vision/datasets/classification/mnist.py:262]: it can either cpu or gpu and we are fine with it (it simulates a user dataset)

File: deepchecks/vision/datasets/classification/mnist.py (hunk 5 above)
Comment (Nadav-Barak) [reply deepchecks/vision/datasets/classification/mnist.py:309]: was there originally, just changed " to ' to make pylint happy

Comment (Nadav-Barak) [reply deepchecks/vision/datasets/classification/mnist.py:309]: but removed the finally anyway

File: deepchecks/vision/context.py (hunk 3 above)
Comment (JKL98ISR) [reply deepchecks/vision/context.py:51]: sad

File: deepchecks/vision/datasets/classification/mnist.py (hunk 6 above)
Comment (JKL98ISR) [reply deepchecks/vision/datasets/classification/mnist.py:305]: it's a quick fix I trust in you, just do logger.warning

File: deepchecks/vision/datasets/classification/mnist.py (hunk 4 above)
Comment (Nadav-Barak) [reply deepchecks/vision/datasets/classification/mnist.py:265]: ok

File: deepchecks/vision/datasets/classification/mnist.py (hunk 6 above)
Comment (Nadav-Barak) [reply deepchecks/vision/datasets/classification/mnist.py:308]: ok

**Review by JKL98ISR — APPROVED** [review]

File: .github/workflows/build.yml [hunk 7]
```diff
@@ -145,7 +157,7 @@ jobs:
       - name: Set Up Env
         run: make env
       - name: Run Tests
-        run: make test
+        run: make test args=tests/vision/gpu_tests
```
Comment (matanper) [inline .github/workflows/build.yml:160]: why did you add to make `test-tabular-only` if you can do it with args like here?

Comment (Nadav-Barak) [reply .github/workflows/build.yml:160]: main reason is it is not installing vision

**Review by ItayGabbay — COMMENTED** [review]
So, we don't support vision on Python 3.6?
That's a good reason to separate the package into namespaced packages.

**Comment by Nadav-Barak** [comment]
> So, we don't support vision on Python 3.6? That's a good reason to separate the package into namespaced packages.
> 
> Did you specify this limitation in the docs?

Agree in concept but we should discuss it further (I will follow up on that)
We will do the doc next sprint, there will be a major refactor in them

File: deepchecks/vision/__init__.py [hunk 8]
```diff
@@ -9,16 +9,23 @@
 # ----------------------------------------------------------------------------
 #
 """Package for vision functionality."""
+try:
+    import torch  # noqa: F401
+    import torchvision  # noqa: F401
+except ImportError as error:
+    raise ImportError('PyTorch is not installed. Please install torch and torchvision '
```
Comment (noamzbr) [inline deepchecks/vision/__init__.py:16]: Does ignite also use torchvision?

File: deepchecks/vision/checks/data_integrity/abstract_property_outliers.py [hunk 9]
```diff
@@ -215,47 +214,61 @@ def get_default_properties(self, data: VisionData):
...
-                                               property_name: str):
+    def _cache_property_values_and_images(self, images: t.List, labels: t.List, property_values: t.List,
+                                          property_name: str):
         """Update the _lowest_property_value_images, _lowest_property_value_images dicts based on new batch."""
+        is_property_per_label = isinstance(property_values[0], (np.ndarray, t.Sequence))
+        # Update full property values cache for outlier calculation
+        if is_property_per_label:
+            self._properties_results[property_name].extend(property_values)
+        else:
+            self._properties_results[property_name].extend([[x] for x in property_values])
```
Comment (noamzbr) [inline deepchecks/vision/checks/data_integrity/abstract_property_outliers.py:225]: What is self._properties_results actually used for?

File: deepchecks/vision/datasets/classification/mnist.py [hunk 10]
```diff
@@ -52,18 +62,21 @@ def load_dataset(
         Train or Test dataset
     batch_size: int, optional
         how many samples per batch to load
-    shuffle : bool , default : False
-        to reshuffled data at every epoch or not
+    shuffle : bool , default : ``False``
```
Comment (noamzbr) [inline deepchecks/vision/datasets/classification/mnist.py:65]: why the `` only here?

File: deepchecks/vision/datasets/classification/mnist.py [hunk 11]
```diff
@@ -52,18 +62,21 @@ def load_dataset(
...
+        to reshuffled data at every epoch or not, cannot work with use_iterable_dataset=True
```
Comment (noamzbr) [inline deepchecks/vision/datasets/classification/mnist.py:66]:
```suggestion
        to reshuffled data at every epoch or not, cannot be used when use_iterable_dataset=True
```

File: deepchecks/vision/datasets/classification/mnist.py [hunk 12]
```diff
@@ -181,33 +195,134 @@ def load_model(pretrained: bool = True, path: pathlib.Path = None) -> 'MockMnist
...
+                prediction = self.real_model(torch.stack([img]).to(self.device))[0]
                 prediction = nn.Softmax()(prediction).detach()
-                self.cache[hash_key] = prediction
+                self.cache[hash_key] = prediction.to(self.device)
             results.append(self.cache[hash_key])
         return torch.stack(results)
 
 
-class MNIST(datasets.MNIST):
+class IterableTorchMnistDataset(IterableDataset):
```
Comment (noamzbr) [inline deepchecks/vision/datasets/classification/mnist.py:223]: Can't this also subclass Mnist but be iterable? Has to download a different copy of mnist by itself?

File: deepchecks/vision/datasets/detection/coco_tensorflow.py [hunk 13]
```diff
@@ -0,0 +1,122 @@
...
+    if shuffle:
+        coco_dataset = coco_dataset.shuffle(128)
+
+    if object_type == 'Dataset':
+        return coco_dataset
+    elif object_type == 'VisionData':
+        model = hub.load(_MODEL_URL)
+        coco_dataset = coco_dataset.map(deepchecks_map(model))
+        return VisionData(batch_loader=coco_dataset, label_map=LABEL_MAP, task_type='object_detection',
+                          shuffle_batch_loader=False)
```
Comment (noamzbr) [inline deepchecks/vision/datasets/detection/coco_tensorflow.py:75]: What happens if user sets that to True for TF?

File: deepchecks/vision/vision_data/vision_data.py [hunk 14]
```diff
@@ -52,6 +52,9 @@ def __init__(
             dataset_name: t.Optional[str] = None,
             shuffle_batch_loader: bool = True
     ):
+        if not hasattr(batch_loader, '__iter__'):
+            raise DeepchecksValueError(r'Batch loader must be an iterable which loads batches of data in deepcheck\'s'
+                                       'required format, see link for additional information ')
```
Comment (noamzbr) [inline deepchecks/vision/vision_data/vision_data.py:57]: Which link? Leave todo

**Review by noamzbr — APPROVED** [review]
Looks good, consider sprinkling in some TF tests in some checks

File: deepchecks/vision/checks/data_integrity/abstract_property_outliers.py (hunk 9 above)
Comment (Nadav-Barak) [reply deepchecks/vision/checks/data_integrity/abstract_property_outliers.py:225]: the actual outlier calculation (stores all values then calculate based on them)

File: deepchecks/vision/__init__.py (hunk 8 above)
Comment (Nadav-Barak) [reply deepchecks/vision/__init__.py:16]: good catch, it does not

Comment (Nadav-Barak) [reply deepchecks/vision/__init__.py:16]: modified

File: deepchecks/vision/datasets/classification/mnist.py (hunk 10 above)
Comment (Nadav-Barak) [reply deepchecks/vision/datasets/classification/mnist.py:65]: removed

File: deepchecks/vision/datasets/classification/mnist.py (hunk 12 above)
Comment (Nadav-Barak) [reply deepchecks/vision/datasets/classification/mnist.py:223]: it downloads uses same copy

File: deepchecks/vision/datasets/detection/coco_tensorflow.py (hunk 13 above)
Comment (Nadav-Barak) [reply deepchecks/vision/datasets/detection/coco_tensorflow.py:75]: Didnt understand

File: deepchecks/vision/vision_data/vision_data.py (hunk 14 above)
Comment (Nadav-Barak) [reply deepchecks/vision/vision_data/vision_data.py:57]: ok

**Comment by Nadav-Barak** [comment]
> Looks good, consider sprinkling in some TF tests in some checks

Lets do that as part of the checks docs refactor, guess for some of them we will use tf dataset
